File: java-algo/src/main/java/com/leetcode/Problem63.java
package com.leetcode;

/**
 * 63. Unique Paths II
 * https://leetcode.com/problems/unique-paths-ii/
 *
 *
 * A robot is located at the top-left corner of a m x n grid
 * (marked 'Start' in the diagram below).
 *
 * The robot can only move either down or right at any point in time.
 * The robot is trying to reach the bottom-right corner of the grid
 * (marked 'Finish' in the diagram below).
 *
 * Now consider if some obstacles are added to the grids.
 * How many unique paths would there be?
 *
 *
 *
 * An obstacle and empty space is marked as 1 and 0 respectively in the grid.
 *
 * Note: m and n will be at most 100.
 *
 * Example 1:
 *
 * Input:
 * [
 *   [0,0,0],
 *   [0,1,0],
 *   [0,0,0]
 * ]
 * Output: 2
 * Explanation:
 * There is one obstacle in the middle of the 3x3 grid above.
 * There are two ways to reach the bottom-right corner:
 * 1. Right -> Right -> Down -> Down
 * 2. Down -> Down -> Right -> Right
 */
public class Problem63 {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int rows = obstacleGrid.length, cols = obstacleGrid[0].length;
        // dp[i][j] means how many roads from start to grid[i][j]
        int[][] dp = new int[rows][cols];
        dp[0][0] = obstacleGrid[0][0] == 0 ? 1 : 0;
        for (int i = 1; i < cols; i++) {
            if (obstacleGrid[0][i] == 0 )
                dp[0][i] = dp[0][i - 1];
        }
        for (int i = 1; i < rows; i++) {
            if (obstacleGrid[i][0] == 0)
                dp[i][0] = dp[i - 1][0];
        }
        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                if (obstacleGrid[i][j] == 0)
                    dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[rows - 1][cols - 1];
    }

    private static final int[][] DIRS = new int[][]{{0, 1}, {1, 0}};
    private static int[][] memo;

    public int uniquePathsWithObstacles2(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        if (grid[0][0] != 0 || grid[m - 1][n - 1] != 0) {
            return 0;
        }
        memo = new int[m][n];
        return dfs(grid, 0, 0);
    }

    private int dfs(int[][] grid, int x, int y) {
        if (memo[x][y] > 0) {
            return memo[x][y];
        }
        if (x == grid.length - 1 && y == grid[0].length - 1) {
            return 1;
        }
        int cnt = 0;
        for (int[] d : DIRS) {
            int nx = x + d[0], ny = y + d[1];
            if (0 <= nx && nx < grid.length && 0 <= ny && ny < grid[0].length && grid[nx][ny] == 0) {
                cnt += dfs(grid, nx, ny);
            }
        }
        return memo[x][y] = cnt;
    }
}

File: python-algo/leetcode/problem_63.py
"""63. Unique Paths II
https://leetcode.com/problems/unique-paths-ii/
"""
from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        dp = [[0] * n for _ in range(m)]
        dp[0][0] = 1
        for i in range(m):
            for j in range(n):
                if obstacleGrid[i][j] == 1:
                    dp[i][j] = 0
                    continue
                dp[i][j] += dp[i - 1][j] if i >= 1 else 0
                dp[i][j] += dp[i][j - 1] if j >= 1 else 0
        return dp[-1][-1]

    def unique_paths_with_obstacles(self, obstacle_grid: List[List[int]]):
        rows, cols = len(obstacle_grid), len(obstacle_grid[0])
        dp = [[1] * cols for _ in range(rows)]
        for _ in range(cols):
            if obstacle_grid[0][_] == 1:
                dp[0][_] = 0
            else:
                dp[0][_] = dp[0][_ - 1]
        for _ in range(rows):
            if obstacle_grid[_][0] == 1:
                dp[_][0] = 0
            else:
                dp[_][0] = dp[_ - 1][0]
        for i in range(1, rows):
            for j in range(1, cols):
                if obstacle_grid[i][j] == 1:
                    dp[i][j] = 0
                else:
                    dp[i][j] = dp[i][j - 1] + dp[i - 1][j]
        return dp[rows - 1][cols - 1]

    def uniquePathsWithObstacles2(self, obstacleGrid: List[List[int]]) -> int:
        if not obstacleGrid or not obstacleGrid[0]:
            return 0
        dp = [1] + [0] * (len(obstacleGrid[0]) - 1)
        for row in obstacleGrid:
            dp[0] = 0 if row[0] == 1 else dp[0]
            for i in range(1, len(obstacleGrid[0])):
                dp[i] = 0 if row[i] == 1 else dp[i] + dp[i - 1]
        return dp[-1]


if __name__ == '__main__':
    sol = Solution()
    print(sol.uniquePathsWithObstacles2([[0, 0, 0], [0, 1, 0], [0, 0, 0]]))
